### ISSUE TITLE
Custom validation - checking variable against object property in array of objects

### ISSUE THREAD

**Issue by rconstantine** [monterail/vuelidate#348]
I've read everything I can on custom validators but still can't figure it out. I don't have any code to show as it all seemed like garbage.

I have a simple string variable in my data section. I need to make sure that it doesn't have a match in an array of objects where the objects look like this:
```javascript
{
  'id': 1,
  'name': 'Management',
  'active': true
}
```

I could simply do the validation myself using something like this:
```javascript
function roleNameIsMatch (theArray, roleName) {
  for (let role in theArray) {
    if (role.name === roleName) {
      return role.id
    }
  }
}
```
...and then fire off all of the same things I'm doing when validation fails, like disabling a save button, but I was hoping to reuse this and understand how to make more validators.

It's possible that the solution is obvious and I'm just an idiot. TIA

**Comment by downace** [comment]
You can try something like this:

```javascript
// Validator
const inArray = (arrFieldName, matches = (e, v) => e === v) => function (value, parentVm) {
    return helpers.ref(arrFieldName, this, parentVm).some(el => matches(el, value));
};

// Usage
new Vue({
    data:        {
        text1: 'Cookie',
        arr1:  [ 'First', 'Second' ],
        
        text2: 'Management',
        arr2:  [
            {
                id:     1,
                name:   'Management',
                active: true
            },
            {
                id:     2,
                name:   'Not Management',
                active: true
            }
        ],
        
        text3: 'Mouse',
        arr3:  [ 'Keyboard', 'Monitor' ],
    },
    validations: {
        text1: {
            // Simple match
            inArray1: inArray('arr1')
        },
        text2: {
            // Custom matcher
            inArray: inArray('arr2', ({ name }, value) => name === value)
        },
        text3: {
            // Inverse match (not is a built-in validator)
            inArray: not(inArray('arr3'))
        }
    }
})
```

Live example: https://jsfiddle.net/downace/Lrj0w5u7/5/

**Comment by rconstantine** [comment]
I will give this a shot. Thanks for the reply. I'll report back after I try it.

**Comment by rconstantine** [comment]
This is working perfectly when I'm adding new items:
const notInArray = (arrFieldName, matches = (e, v) => e === v) => function (value, parentVm) {
  return !helpers.ref(arrFieldName, this, parentVm).some(el => matches(el, value))
}

I had to invert the boolean result was all. One issue is that it doesn't work when I'm editing a row, because it matches itself. I'll have to figure out how to pass a row ID. I suppose I should look at how to incorporate the use of 'param'. Looks straight forward, but I have to put this down for the day and get some other work done. Will try it tomorrow.

Many thanks for the help.

**Comment by rconstantine** [comment]
I just wanted to thank downace again and post the rest of my solution for others.

Like I said above, when I edit existing rows, a match was always found so I needed to fix that. I originally thought I needed to understand Vuelidate a bit more and tried to grok params, but it turned out I needed to understand Javascript a bit more. I only recently returned to js after about 10 years of not using it. But the solution was easy.

I have an element in my Vue data() block called roleSelected which gets set when one clicks on the edit button for a given row in a table, opening a popup dialog. Alternatively, I've also managed to get inline editing working, so one could also click on the actual text element in the row, which changes to an input field. roleSelected is also set in that case.

In either case, hitting enter saves the data only if there are no validation errors, and hitting escape resets the value to the original value. The popup also has buttons for save and cancel.

Anyway, you'll notice that all I had to do to accommodate editing of existing values, and still have it work for new values, was to add the index to the 'some' callback, then check it against the roleSelected value in the parentVm object.

```javascript
const notInArray = (arrFieldName, matches = (e, v) => e === v, index) => function (value, parentVm) {
  return !helpers.ref(arrFieldName, this, parentVm).some((el, idx) => {
    return matches(el, value) && idx !== parentVm.roleSelected
  })
}
```
